### PR TITLE
Use SAS tokens to get an identity in Connectivity tests

### PR DESCRIPTION
**Pull request by gauravIoTEdge**
This PR introduces the use of short-lived SAS tokens to get an identity using the AzureCLI task. The use of shared keys to access the blob storage account is now discouraged, prompting this change.

We don't always upload logs to the storage account, only when we want to debug something. However, log upload is currently broken after the shared key on the storage account was disabled.

This is the manually triggered test run for this PR:

image

- [X] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).

### CODE REVIEW

**Review by damonbarry — APPROVED** [review]

File: builds/e2e/connectivity.yaml [hunk 1]
```diff
@@ -169,6 +168,37 @@ jobs:
...
+          inlineScript: |
+            # Hardcoded resource group and storage account name
+            resource_group="iotedge-deploy"
+            storage_account="iotedgeconnectivitytest"
+          
+            # Set container name with current date
+            container_name="logs$(date +%Y%m%d)"
+          
+            # Set expiry time for the SAS token
+            expiry_time="$(date --utc --date "+2 days" +"%Y-%m-%dT%H:%MZ")"
```
Comment (yophilav) [inline builds/e2e/connectivity.yaml:186]: Let's make this a 12h token than 2 days please.

File: builds/e2e/connectivity.yaml [hunk 2]
```diff
@@ -169,6 +168,37 @@ jobs:
...
+          
+            # Generate the SAS token for the container
+            sasToken="$(az storage container generate-sas --account-name "$storage_account" --name "$container_name" --https-only --expiry "$expiry_time" --as-user --auth-mode login --permissions drwl --output tsv)"
+            if [ $? -ne 0 ]; then
+              echo "Failed to generate SAS token."
```
Comment (yophilav) [inline builds/e2e/connectivity.yaml:191]: You can still print the `$sasToken` for the error message here

File: builds/e2e/connectivity.yaml [hunk 3]
```diff
@@ -169,6 +168,37 @@ jobs:
...
+            storage_account="iotedgeconnectivitytest"
+          
+            # Set container name with current date
+            container_name="logs$(date +%Y%m%d)"
+          
+            # Set expiry time for the SAS token
+            expiry_time="$(date --utc --date "+2 days" +"%Y-%m-%dT%H:%MZ")"
+          
+            # Generate the SAS token for the container
+            sasToken="$(az storage container generate-sas --account-name "$storage_account" --name "$container_name" --https-only --expiry "$expiry_time" --as-user --auth-mode login --permissions drwl --output tsv)"
```
[the remark this answers is not in the document]
Comment (gauravIoTEdge) [reply builds/e2e/connectivity.yaml:189]: Well, I took the inline script and ran it on my machine and visually inspected the SAS token (in addition to the

File: builds/e2e/connectivity.yaml [hunk 4]
```diff
@@ -169,6 +168,37 @@ jobs:
...
+      - task: AzureCLI@2
+        displayName: 'Get SAS URI for Blob Storage Account'
+        inputs:
+          azureSubscription: $(azure.subscription)
+          scriptType: 'bash'
+          scriptLocation: 'inlineScript'
+          inlineScript: |
+            # Hardcoded resource group and storage account name
+            resource_group="iotedge-deploy"
+            storage_account="iotedgeconnectivitytest"
```
[the remark this answers is not in the document]
Comment (gauravIoTEdge) [reply builds/e2e/connectivity.yaml:180]: It's an inline script with no external dependencies. We would be adding an external dependency to a self-contained code block for no reason. connectivity.yaml is not a template, it's just used for connectivity runs. We are unlikely to change the name of the storage accounts.

I would argue that the change that (eventually) should happen is to take all these inline scripts -- the same code is and will be duplicated in multiple places -- and put them into a single script for easier maintenance. And add pipeline variables to that change.

But let's not worry about that right now please. That's an easy change to make later, once everything actually works (the E2E tests work with this bit of code -- they write to the storage account and everything BUT for some reason, the same code won't write to the storage account for connectivity -- maybe it's something simple, maybe not.. I don't know yet).

Once that common code has been pulled out, we would have no choice but to add pipeline variables (so you're right). Please let it slide for now till we get stuff working.

File: scripts/linux/trcE2ETest.sh [hunk 5]
```diff
@@ -917,7 +917,7 @@ function configure_connectivity_settings() {
 
     TEST_INFO="$TEST_INFO,TestDuration=${TEST_DURATION}"
 
-    log_upload_enabled=false
+    log_upload_enabled=true
```
Comment (damonbarry) [inline scripts/linux/trcE2ETest.sh:920]: Should this variable get set back to false before merging the PR?

Comment (gauravIoTEdge) [reply scripts/linux/trcE2ETest.sh:920]: Sure, I'll do it. 

It's much of a muchness at this point though, the upload will still fail, though I guess it's worth not waiting around for it to try and upload.

**Review by yophilav — APPROVED** [review]

**Review by damonbarry — APPROVED** [review]

**Pull request merged**